mpc-core/src/protocols/rep3_ring.rs

```diff
@@ -2,16 +2,107 @@
 //!
 //! This module implements the rep3 share and combine operations for rings
 
+use rand::{distributions::Standard, prelude::Distribution, CryptoRng, Rng};
+use ring::{int_ring::IntRing2k, ring_impl::RingElement};
+
 pub mod arithmetic;
 pub mod binary;
+pub mod casts;
 pub mod conversion;
 mod detail;
-pub(crate) mod ring;
-
-use ring::bit::Bit;
+pub mod ring;
+pub mod yao;
 
 /// Shorthand type for a secret shared bit.
-pub type Rep3BitShare = Rep3RingShare<Bit>;
-
+pub type Rep3BitShare = Rep3RingShare<ring::bit::Bit>;
 pub use arithmetic::types::Rep3RingShare;
-pub mod yao;
+
+/// Secret shares a ring element using replicated secret sharing and the provided random number generator. The ring element is split into three additive shares, where each party holds two. The outputs are of type [`Rep3RingShare`].
+pub fn share_ring_element<T: IntRing2k, R: Rng + CryptoRng>(
+    val: RingElement<T>,
+    rng: &mut R,
+) -> [Rep3RingShare<T>; 3]
+where
+    Standard: Distribution<T>,
+{
+    let a = rng.gen::<RingElement<T>>();
+    let b = rng.gen::<RingElement<T>>();
+    let c = val - a - b;
+    let share1 = Rep3RingShare::new_ring(a, c);
+    let share2 = Rep3RingShare::new_ring(b, a);
+    let share3 = Rep3RingShare::new_ring(c, b);
+    [share1, share2, share3]
+}
+
+/// Secret shares a vector of ring elements using replicated secret sharing and the provided random number generator. The ring elements are split into three additive shares each, where each party holds two. The outputs are of type [`Rep3RingShare`].
+pub fn share_ring_elements<T: IntRing2k, R: Rng + CryptoRng>(
+    vals: &[RingElement<T>],
+    rng: &mut R,
+) -> [Vec<Rep3RingShare<T>>; 3]
+where
+    Standard: Distribution<T>,
+{
+    let mut shares1 = Vec::with_capacity(vals.len());
+    let mut shares2 = Vec::with_capacity(vals.len());
+    let mut shares3 = Vec::with_capacity(vals.len());
+    for val in vals {
+        let [share1, share2, share3] = share_ring_element(val.to_owned(), rng);
+        shares1.push(share1);
+        shares2.push(share2);
+        shares3.push(share3);
+    }
+    [shares1, shares2, shares3]
+}
+
+/// Secret shares a ring element using replicated secret sharing and the provided random number generator. The ring element is split into three binary shares, where each party holds two. The outputs are of type [`Rep3RingShare`].
+pub fn share_ring_element_binary<T: IntRing2k, R: Rng + CryptoRng>(
+    val: RingElement<T>,
+    rng: &mut R,
+) -> [Rep3RingShare<T>; 3]
+where
+    Standard: Distribution<T>,
+{
+    let a = rng.gen::<RingElement<T>>();
+    let b = rng.gen::<RingElement<T>>();
+    let c = val ^ a ^ b;
+    let share1 = Rep3RingShare::new_ring(a, c);
+    let share2 = Rep3RingShare::new_ring(b, a);
+    let share3 = Rep3RingShare::new_ring(c, b);
+    [share1, share2, share3]
+}
+
+//TODO RENAME ME TO COMBINE_ARITHMETIC_SHARE
```

Comment:
do we do this rename here?

Reply:
This is copied from the original rep3, don't know who added that comment...

Reply:
I removed it